api/admin/controller/collection_settings.py

```diff
@@ -0,0 +1,266 @@
+from api.enki import EnkiAPI
+from api.odl import ODLWithConsolidatedCopiesAPI, SharedODLAPI
+from core.model import (
+    Collection,
+    ConfigurationSetting,
+    ExternalIntegration,
+    get_one,
+    get_one_or_create,
+    Library,
+)
+from core.util.problem_detail import ProblemDetail
+
+class CollectionSettingsController(SettingsController):
+
+    # SET UP
+    def process_collections(self):
+        self.set_up_protocols()
+        if flask.request.method == 'GET':
+            return self.process_get()
+        else:
+            return self.process_post()
+
+    def set_up_protocols(self):
+        self.protocols = self._get_collection_protocols(self.PROVIDER_APIS)
+
+        # If there are storage integrations, add a mirror integration
+        # setting to every protocol's 'settings' block.
+        mirror_integration_setting = self._mirror_integration_setting()
+        if mirror_integration_setting:
+            for protocol in self.protocols:
+                protocol['settings'].append(mirror_integration_setting)
+
+    # GET
+    def process_get(self):
+        user = flask.request.admin
+        collections = []
+        protocolClass = None
+        for collection_object in self._db.query(Collection).order_by(Collection.name).all():
+            if not user or not user.can_see_collection(collection_object):
+                continue
+
+            collection_dict = self.collection_to_dict(collection_object)
+
+            if collection_object.protocol in [p.get("name") for p in self.protocols]:
+                [protocol] = [p for p in self.protocols if p.get("name") == collection_object.protocol]
+                self.load_libraries(collection_dict, collection_object, user, protocol)
+                self.load_settings(protocol.get("settings"), collection_dict, collection_object)
+                [protocolClass] = self.find_protocol_class(collection_object)
+
+            collection_dict["self_test_results"] = self._get_prior_test_results(collection_object, protocolClass)
+            collections.append(collection_dict)
+
+        return dict(
+            collections=collections,
+            protocols=self.protocols,
+        )
+
+    def collection_to_dict(self, collection_object):
+        return dict(
+            id=collection_object.id,
+            name=collection_object.name,
+            protocol=collection_object.protocol,
+            parent_id=collection_object.parent_id,
+            settings=dict(external_account_id=collection_object.external_account_id),
+        )
+
+    def load_libraries(self, collection_dict, collection_object, user, protocol):
+        libraries = []
+        for library in collection_object.libraries:
+            if not user or not user.is_librarian(library):
+                continue
+            libraries.append(self._get_integration_library_info(
+                    collection_object.external_integration, library, protocol))
+
+        collection_dict['libraries'] = libraries
```

Comment:
This is a really minor point, but this method would be more reusable if it didn't take `collection_dict` as an argument and it just returned `libraries`. The caller would be responsible for deciding what to do with `libraries`.